src/proxy/cpsubsystem/SessionAwareSemaphoreProxy.ts

```diff
@@ -0,0 +1,302 @@
+    }
+
+    release(permits = 1): Promise<void> {
+        assertNonNegativeNumber(permits);
+
+        const sessionId = this.getSessionId();
+        if (NO_SESSION_ID.equals(sessionId)) {
+            return Promise.reject(this.newIllegalStateError());
+        }
+        const threadId = Long.fromNumber(this.nextThreadId());
+        const invocationUid = UuidUtil.generate();
+        return this.requestRelease(sessionId, threadId, invocationUid, permits)
+            .catch((err) => {
+                if (err instanceof SessionExpiredError) {
+                    this.invalidateSession(sessionId);
+                    throw this.newIllegalStateError(err);
+                }
+                this.releaseSession(sessionId, permits);
+                throw err;
+            })
+            .then(() => {
+                this.releaseSession(sessionId, permits);
+            });
+    }
+
+    availablePermits(): Promise<number> {
+        return this.encodeInvokeOnRandomTarget(SemaphoreAvailablePermitsCodec, this.groupId, this.objectName)
+            .then((clientMessage) => {
+                const response = SemaphoreAvailablePermitsCodec.decodeResponse(clientMessage);
+                return response.response;
+            });
+    }
+
+    drainPermits(): Promise<number> {
+        const invocationUid = UuidUtil.generate();
+        return this.doDrainPermits(invocationUid);
+    }
+
+    private doDrainPermits(invocationUid: UUID): Promise<number> {
+        let sessionId: Long;
+        const threadId = Long.fromNumber(this.nextThreadId());
+        return this.acquireSession(DRAIN_SESSION_ACQ_COUNT)
+            .then((id) => {
+                sessionId = id;
+                return this.requestDrain(sessionId, threadId, invocationUid);
+            })
+            .then((count) => {
+                this.releaseSession(sessionId, DRAIN_SESSION_ACQ_COUNT - count);
+                return count;
+            })
+            .catch((err) => {
+                if (err instanceof SessionExpiredError) {
+                    this.invalidateSession(sessionId);
+                    return this.doDrainPermits(invocationUid);
+                }
+                throw err;
+            });
+    }
+
+    reducePermits(reduction: number): Promise<void> {
+        assertNonNegativeNumber(reduction);
+        if (reduction === 0) {
+            return;
+        }
+        return this.doChangePermits(-reduction);
+    }
+
+    increasePermits(increase: number): Promise<void> {
+        assertNonNegativeNumber(increase);
+        if (increase === 0) {
+            return;
+        }
+        return this.doChangePermits(increase);
+    }
+
+    private doChangePermits(delta: number): Promise<void> {
+        let sessionId: Long;
+        const threadId = Long.fromNumber(this.nextThreadId());
+        const invocationUid = UuidUtil.generate();
+        return this.acquireSession()
+            .then((id) => {
+                sessionId = id;
```

Comment:
That check is redundant. I see no way for `AbstractProxySessionManager#acquireSession` to return `NO_SESSION_ID`. Also, other methods that also call `acquireSession` lack such check.

Do you see any scenarios when it might be returned? If not, I'll submit a cleanup PR for Java client.

Reply:
I checked the server side and I guess you are right. I didn't find any codepaths that might return this

Reply:
Here comes the cleanup PR: https://github.com/hazelcast/hazelcast/pull/17540